### PR TITLE
Feature/hint compat api css

### PR DESCRIPTION
**Pull request by CKGrafico**
---------
This PR is in development, check first comment
---------
This PR has to implement 2 of the multiple hints about Compat Api everything is detailed in https://github.com/webhintio/rfcs/issues/10

But I'm moving the checklist to this PR:
To do for CSS part of this task:
* [x] Create: Multi hint named `hint-compat-api` and organize it by language (CSS, JS, HTML)
* [x] Create: Also split each of this by deprecated and not broadly supported
* [X] Feature: Which properties or values are deprecated in the CSS
    * [x] To Do: Take Browsers data from config.
    * [x] To Do: Take MDN data for the API to check (CSS) (Generic service).
    * [x] To Do: Merge this two sources to have one source of truth.
    * [x] To Do: For Each parsed key (.walk()) check if is suported or not.
    * [x] To Do: Report if there is an error
    * [x] To Do:  Add tests
    * [x] To Do:  Docs
* [x] Feature: Which properties or values are not broadly supported in the CSS.
    * [X] To Do: For Each parsed key (.walk()) check if is not broadly supported.
    * [X] To Do: Report if there is an error
    * [X] To Do: Add tests
    * [x]  To Do: Docs

* [x] Feature: Sometimes features are multilevel and they should work.
* [x] Feature: If the user had user prefixes for one browser and feature has to not report erros in not broadly supported properties.

* [x] Check: Avoid to throw errors if the user configured to do not support a specified browser (ex: API is deprecated for IE but the user configured hint only for Chrome) (Maybe we need more info to be sure about this point)

* [x] To Do: Use Location into reports
* [x] To Do: Don't check more than once the same scenario (cache)
* [x] To Do: Check foreach loops

* [x] Feedback: Add location problem to tests

* [x] Review: All relevant code is documented
* [x] Review: All the functionality is covered by tests and is near 100% of coverage
* [x] Review: All the functionality is documented in docs
* [x] Review: Check with some real websites

Bugs or questions for 3rd party:
* [x] https://github.com/mdn/browser-compat-data/issues/3034
* [x] https://github.com/webhintio/rfcs/issues/10#issuecomment-433424902

### CODE REVIEW

**Comment by CKGrafico** [comment]
This PR is in development but feel free to give feedback about this files:
 - 
 -
More info in https://github.com/webhintio/rfcs/issues/10 (structure and details)

**Comment by molant** [comment]
@antross I believe this is ready for a first pass of feedback if you have time.

File: packages/hint-compat-api/src/helpers/compat-api.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                namespaceFeatures[featureKey] = typedFeatureValue;
+            });
+
+            compatDataApi[namespaceFeaturesKey] = namespaceFeatures;
+        });
+
+        return compatDataApi;
+    }
+
+    public getPrefix(name: string): [string, string] | [undefined, string] {
```
Comment (antross) [inline packages/hint-compat-api/src/helpers/compat-api.ts]: Slightly more concise:
```suggestion
    public getPrefix(name: string): [string | undefined, string] {
```

File: packages/hint-compat-api/src/compat-api-css.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        id: 'compat-api-css',
+        schema: [],
+        scope: HintScope.any
+    }
+
+    public constructor(context: HintContext) {
+        const onParseCSS = (styleParse: StyleParse): void => {
+            const { resource } = styleParse;
+            const mdnBrowsersCollection = userBrowsers.convert(context.targetedBrowsers);
+            const compatApi = new CompatApi('css', mdnBrowsersCollection);
```
[the remark this answers is not in the document]
Comment (CKGrafico) [reply packages/hint-compat-api/src/compat-api-css.ts]: You're perfectly right! everything should be outside onParse

File: packages/hint-compat-api/src/helpers/compat-api.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    public getPrefix(name: string): [string | undefined, string] {
+        const regexp = new RegExp(`-(moz|o|webkit|ms)-`, 'gi');
+        const matched = name.match(regexp);
+
+        return matched && matched.length > 0 ? [matched[0], name.replace(matched[0], '')] : [undefined, name];
+    }
+
+    public getSupportStatementFromInfo(browserFeatureSupported?: SupportStatement, prefix?: string): SimpleSupportStatement | undefined {
```
Comment (molant) [inline packages/hint-compat-api/src/helpers/compat-api.ts]: I don't think we are memoizing the result of anything (PR is big so I might have missed it) and I imagine we will be checking multiple times the same properties. If that's the case we should look into adding that.

File: packages/hint-compat-api/src/helpers/compat-api.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        return browserFeatureSupported as SimpleSupportStatement;
+    }
+    /* eslint-enable camelcase */
+
+    private isFeatureRequiredToTest(typedFeatureValue: CompatStatement & MDNTreeFilteredByBrowsers, isCheckingNotBroadlySupported = false): boolean {
+        // TODO: Here we are checking only parent but this object has children
+        let isRequiredToTest = false;
+
+        forEach(this.browsers, (browserVersionsList, browser) => {
+            if (isRequiredToTest || !typedFeatureValue.__compat || !typedFeatureValue.__compat.support) {
```
Comment (molant) [inline packages/hint-compat-api/src/helpers/compat-api.ts]: I'd do something similar to what I proposed in my previous comment about using `every` to validate everything so it's easier to know what we are validating.

File: packages/hint-compat-api/src/helpers/compat-api.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            const { version_added: addedVersion, version_removed: removedVersion } = browserFeatureSupported;
+
+            if (isCheckingNotBroadlySupported) {
+                if (!addedVersion || isNaN(parseFloat(addedVersion.toString()))) {
+                    return;
+                }
+
+                if (addedVersion || browserVersionsList[0] <= browserVersions.normalize(addedVersion)) {
+                    isRequiredToTest = true;
+                }
```
Comment (molant) [inline packages/hint-compat-api/src/helpers/compat-api.ts]: Maybe add a `return` here so we can remove the else in line 143 and have the next `if`s with less nesting?

File: packages/hint-compat-api/src/helpers/compat-api.ts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+        return browserFeatureSupported as SimpleSupportStatement;
+    }
+    /* eslint-enable camelcase */
+
+    private isFeatureRequiredToTest(typedFeatureValue: CompatStatement & MDNTreeFilteredByBrowsers, isCheckingNotBroadlySupported = false): boolean {
+        // TODO: Here we are checking only parent but this object has children
+        let isRequiredToTest = false;
+
+        forEach(this.browsers, (browserVersionsList, browser) => {
```
Comment (molant) [inline packages/hint-compat-api/src/helpers/compat-api.ts]: If I've understood the code correctly, this could be improved by doing [`Array.prototype.some()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) so the loop stops the moment there's a `true` value returned.

File: packages/hint-compat-api/src/compat-api-css-next.ts [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+/*
+ * ------------------------------------------------------------------------------
+ * Public
+ * ------------------------------------------------------------------------------
+ */
+
+export default class implements IHint {
+
+    public static readonly meta: HintMetadata = {
```
Comment (antross) [inline packages/hint-compat-api/src/compat-api-css-next.ts]: Just a heads-up: we're changing how `HintMetadata` is declared to facilitate loading it separate from hints (e.g. for building documentation). See `hint-typescript-config` in PR #1480 for how I'm proposing we do this for multi-hints.

**Comment by molant** [comment]
Nice work so far @CKGrafico and @isabellachen !

**Comment by CKGrafico** [comment]
Added some improvements after check with real websites.
Added new task (To add problem location to the report)
Pending: Check @molant Feedback

File: packages/hint-compat-api/src/compat-api-css-next.ts [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+            category: Category.interoperability,
+            description: `Hint to validate if the CSS features of the project are not broadly supported`
+        },
+        id: 'compat-api-css-next',
+        schema: [],
+        scope: HintScope.any
+    }
+
+    public constructor(context: HintContext) {
+        const mdnBrowsersCollection = userBrowsers.convert(context.targetedBrowsers);
```
[the remark this answers is not in the document]
Comment (CKGrafico) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: @isabellachen can you check this?

File: packages/hint-compat-api/src/helpers/compat-api.ts (hunk 3 above)
Comment (CKGrafico) [reply packages/hint-compat-api/src/helpers/compat-api.ts]: Adding as a task

File: packages/hint-compat-api/src/helpers/compat-api.ts (hunk 6 above)
Comment (CKGrafico) [reply packages/hint-compat-api/src/helpers/compat-api.ts]: Yes! nice feedback, we will use _.some to be easier for collections if this is ok

Comment (molant) [reply packages/hint-compat-api/src/helpers/compat-api.ts]: I'd rather us use the builtin methods. So far the only methods used from Lodash are `forEach` and now `some` too which IMO doesn't justify using it and pulling another package.

**Comment by isabellachen** [comment]
@CKGrafico CompatApi is not taking care of "version_added": "null" for children. It currently omits any feature that is not required to test only up to one level...

**Comment by isabellachen** [comment]
@CKGrafico Another edge case with the mdn compat data: compat data for flex properties, like justify-content: space-evenly are three levels deep as the compat data for space-evenly is stored in a 'flex_context' property under justify-content. I.e. data['justify-content']['flex_context']['space-evenly']

**Comment by CKGrafico** [comment]
Review tasks:

* [ ] Add concrete examples in normalize functions (xx.xx.xx)
* [ ] Test that all the browser names are ok, on browser covnersions
* [ ] Add concrete examples in getWorstCaseSupportStatementFromInfo
* [ ] In compat-api.ts divide bool check and number check in two blocks
* [ ] Show one error for eac cached feature (and new problem location)
* [ ] Dblcheck all the any, maybe some of them could be replaced
* [ ] Check the `undefined`insed compat-api
* [ ] Add concrete examples in getSupportStatementFromInfo
* [ ] Be sure that is not throwing errors in scenarios like:
```
    -moz- : added 21
    '' : added 35
 
    supporting 34-35
```
* [ ] Join error versions like:
```
chrome 32, chrome 33, chrome 34
chrome 32-35
```
* [ ] Use singular if the list of errors has 1 item
* [ ] Array<HintTest>  is HintTest[]

**Comment by CKGrafico** [comment]
@antross @molant 
Today I've discovered that there are some basic css properties like `text-transform`https://developer.mozilla.org/es/docs/Web/CSS/text-transform or `min-height` https://developer.mozilla.org/es/docs/Web/CSS/min-height 

where we are reporting errors like:

```
line 6    col 10  Error  text-transform of CSS is not added on chrome_android browser.                                                       compat-api-css-next
line 7    col 10  Error  text-transform of CSS is not added on samsunginternet_android browser.                                              compat-api-css-next
```
only on mobile browsers, this is happening because in the MDN API there is no info about this browsers.

I think that to report this error is technically correct, but maybe can confuse to the user.
Do you still think that  the best is to do PR to MDN package and try to update their API or do you prefer to do something manually for our package?

**Comment by molant** [comment]
Do a pr to MDN. We want the data coming from there as accurate as possible.
Thanks!


________________________________

**Comment by CKGrafico** [comment]
Merging the feedback from the review @molant.

I decided to move this task apart because has more work than expected.

* [ ] Be sure that is not throwing errors in scenarios like:
```
    -moz- : added 21
    '' : added 35
 
    supporting 34-35
```

After this task the last thing will be autoprefixer ¿Do you want to use the same PR?

File: packages/hint-compat-api/src/compat-api-css-next.ts [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+                }
+            });
+        };
+
+        const compatCSS = new CompatCSS(checkNotBroadlySupportedFeature);
+
+        const onParseCSS = (styleParse: StyleParse): void => {
+            const { resource } = styleParse;
+
+            compatCSS.searchCSSFeatures(compatApi.compatDataApi, mdnBrowsersCollection, styleParse, resource);
```
[the remark this answers is not in the document]
Comment (CKGrafico) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: Yes

File: packages/hint-compat-api/src/compat-api-css-next.ts [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+                    forEach(browsersToSupport, (versions, browserName) => {
+                        if (browserName !== browserToSupportName) {
+                            return;
+                        }
+
+                        wasSupportedInSometime = true;
+                    });
+
+                    if (!wasSupportedInSometime) {
+                        context.report(resource, null, `${featureName} of CSS was never added on ${browserToSupportName} browser.`, featureName);
```
[the remark this answers is not in the document]
Comment (CKGrafico) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: Second time that we forget this.. solved. Maybe we need to add some other async / await but I have not found more.

File: packages/hint-compat-api/src/compat-api-css-next.ts (hunk 7 above)
Comment (CKGrafico) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: Ok thanks

**Comment by molant** [comment]
@CKGrafico we've merged the metadata change. If you can update your PR to follow that I think this should be ready for merge and continue working on the things that you mention in your comment.

@antross, @sarvaje, @alrra thoughts?

File: packages/hint-compat-api/package.json [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    "extends": "../../.nycrc"
+  },
+  "peerDependencies": {
+    "hint": "^3.4.10"
+  },
+  "private": true,
+  "repository": "webhintio/hint",
+  "scripts": {
+    "build": "npm run clean && npm-run-all build:*",
+    "build-release": "npm run clean && npm run build:assets && tsc -b --inlineSourceMap false --removeComments true",
```
Comment (alrra) [inline packages/hint-compat-api/package.json]:
```suggestion
    "build-release": "npm run clean && npm run build:assets && tsc --inlineSourceMap false --removeComments true",
```

Ref: https://github.com/webhintio/hint/pull/1411

File: packages/hint-compat-api/docs/compat-api-css-next.md [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+while targeting Chrome browsers of versions less than 29
+will trigger the hint.
+
+```css
+.example {
+    display: flex;
+}
+```
+
+The at-rule [keyframes](https://developer.mozilla.org/en-US/docs/Web/CSS/@keyframes) was added for versions of Chrome 43 and onwards. Using `@keyframes`
```
[the remark this answers is not in the document]
Comment (alrra) [reply packages/hint-compat-api/docs/compat-api-css-next.md]: Keep the line length under 80 characters.

File: packages/hint-compat-api/docs/compat-api-css.md [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+```css
+.example {
+    box-lines: single;
+}
+```
+
+## Can the hint be configured?
+
+This hint throws errors for CSS features that have been deprecated in any of the [targeted browsers](../../hint/docs/user-guide/configuring-webhint/browser-context.md) listed.
```
[the remark this answers is not in the document]
Comment (alrra) [reply packages/hint-compat-api/docs/compat-api-css.md]: Keep the line under 80 characters.

File: packages/hint-compat-api/docs/compat-api-css.md [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+The targeted browsers can be defined in either the `.hintrc` or `package.json` file. This property follows the same convention as [browserlist](https://github.com/browserslist/browserslist#readme).
```
[the remark this answers is not in the document]
Comment (alrra) [reply packages/hint-compat-api/docs/compat-api-css.md]: Keep the line under 80 characters.

File: packages/hint-compat-api/src/helpers/compat-api.ts [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+                    return true;
+                }
+            }
+
+            return false;
+        });
+    }
+
+    public isBrowserToSupportPartOfBrowsersCollection(browsersToSupport: BrowserSupportCollection, browserToSupportName: string): boolean {
+        if (!Object.keys(browsersToSupport).some((browser) => {
```
Comment (sarvaje) [inline packages/hint-compat-api/src/helpers/compat-api.ts]: Move the condition to another method, it is a little bit difficult to read.

File: packages/hint-compat-api/src/helpers/compat-css.ts [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+            let groupStarted = false;
+
+            sortedVersions.forEach((value, i) => {
+                const nextValue = sortedVersions[i + 1];
+
+                if (!groupStarted) {
+                    grouped += `${browser} ${value}`;
+                }
+
+                if (nextValue && Number(nextValue) - Number(value) > 1) {
```
Comment (sarvaje) [inline packages/hint-compat-api/src/helpers/compat-css.ts]: value and nextValue is not a number by default?

**Comment by CKGrafico** [comment]
@antross @molant merged with master and updated metadata + events names + context.report changes.

File: packages/hint-compat-api/README.md [hunk 17]
```diff
@@ -0,0 +1,51 @@
+# compat-api (`@hint/hint-compat-api`)
+
+description
+
+## Why is this important?
+
+Explain why this package is important for your users
```
[the remark this answers is not in the document]
Comment (sarvaje) [reply packages/hint-compat-api/README.md]: I have talked to @molant and this is not necessary yet.

File: packages/hint-compat-api/README.md [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+        "compat-api/compat-api-css": "error",
+        "compat-api/compat-api-css-next": "error"
+    },
+    ...
+}
+```
+
+## Further Reading
+
+What can the user read to know more about this subject?
```
[the remark this answers is not in the document]
Comment (sarvaje) [reply packages/hint-compat-api/README.md]: I have talked to @molant and this is not necessary yet.

File: packages/hint-compat-api/README.md [hunk 19]
```diff
@@ -0,0 +1,51 @@
+# compat-api (`@hint/hint-compat-api`)
+
+description
```
[the remark this answers is not in the document]
Comment (CKGrafico) [reply packages/hint-compat-api/README.md]: Modified, inspired in https://github.com/webhintio/hint/edit/master/packages/hint-typescript-config/README.md

**Comment by CKGrafico** [comment]
Applied @alrra feedback about md, thanks

File: packages/hint-compat-api/README.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (CKGrafico) [reply packages/hint-compat-api/README.md]: Solved with @alrra feedback

File: packages/hint-compat-api/src/compat-api-css-next.ts [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+    private compatApi: CompatApi;
+    private compatCSS: CompatCSS;
+    private userPrefixes: UserPrefixes = {};
+
+    public constructor(context: HintContext) {
+        const isCheckingNotBroadlySupported = true;
+
+        this.mdnBrowsersCollection = userBrowsers.convert(context.targetedBrowsers);
+        this.compatApi = new CompatApi('css', this.mdnBrowsersCollection, isCheckingNotBroadlySupported);
+        this.compatCSS = new CompatCSS(context, (...params) => {
```
[the remark this answers is not in the document]
Comment (CKGrafico) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: This is  not going to work :(

File: packages/hint-compat-api/src/compat-api-css-next.ts [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+            this.testFeatureIsSupportedInBrowser(...params);
+        });
+
+        context.on('parse::css::end', async (styleParse: StyleParse) => {
```
[the remark this answers is not in the document]
Comment (CKGrafico) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: Same as before

File: packages/hint-compat-api/src/compat-api-css.ts [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+import { browserVersions } from './helpers/normalize-version';
+
+/*
+ * ------------------------------------------------------------------------------
+ * Public
+ * ------------------------------------------------------------------------------
+ */
+
+export default class implements IHint {
+    public static readonly meta: HintMetadata = {
```
[the remark this answers is not in the document]
Comment (CKGrafico) [reply packages/hint-compat-api/src/compat-api-css.ts]: Done with merge

File: packages/hint-compat-api/src/compat-api-css.ts [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+    private compatCSS: CompatCSS;
+
+    public constructor(context: HintContext) {
+        this.mdnBrowsersCollection = userBrowsers.convert(context.targetedBrowsers);
+        this.compatApi = new CompatApi('css', this.mdnBrowsersCollection);
+        this.compatCSS = new CompatCSS(context, (...params) => {
+            this.testFeatureIsSupportedInBrowser(...params);
+        });
+
+        context.on('parse::css::end', async (styleParse: StyleParse) => {
```
[the remark this answers is not in the document]
Comment (CKGrafico) [reply packages/hint-compat-api/src/compat-api-css.ts]: Same as before

File: packages/hint-compat-api/src/helpers/cached-compat-features.ts [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    public async showCachedErrors(featureName: string, context: HintContext, location?: ProblemLocation): Promise<void> {
+        const cachedErrors = this.cachedFeatures[featureName];
+
+        if (!cachedErrors || cachedErrors.length < 1) {
+            return;
+        }
+
+        await cachedErrors.forEach(async (cachedFeature: CachedFeature) => {
```
[the remark this answers is not in the document]
Comment (CKGrafico) [reply packages/hint-compat-api/src/helpers/cached-compat-features.ts]: True, solved

File: packages/hint-compat-api/src/helpers/compat-api.ts (hunk 15 above)
Comment (CKGrafico) [reply packages/hint-compat-api/src/helpers/compat-api.ts]: I hope is better now:
```
const isBrowserInsideCollection = Object.keys(browsersToSupport).some((browser) => {
            return browser === browserToSupportName;
        })

        return isBrowserInsideCollection;
```

File: packages/hint-compat-api/src/helpers/compat-css.ts (hunk 16 above)
Comment (CKGrafico) [reply packages/hint-compat-api/src/helpers/compat-css.ts]: Can be something like 4.4.4 https://github.com/mdn/browser-compat-data/issues/3034
image

But I've updated to be safe against '4.4.4' because we forgot, thanks.

File: packages/hint-compat-api/src/compat-api-css.ts [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    private mdnBrowsersCollection: BrowserSupportCollection;
+    private compatApi: CompatApi;
+    private compatCSS: CompatCSS;
+
+    public constructor(context: HintContext) {
+        this.mdnBrowsersCollection = userBrowsers.convert(context.targetedBrowsers);
+        this.compatApi = new CompatApi('css', this.mdnBrowsersCollection);
+        this.compatCSS = new CompatCSS(context, (...params) => {
```
[the remark this answers is not in the document]
Comment (CKGrafico) [reply packages/hint-compat-api/src/compat-api-css.ts]: Same as before

**Comment by CKGrafico** [comment]
Applied @sarvaje feedback and added some comments, thanks guys!

File: packages/hint-compat-api/src/compat-api-css-next.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (sarvaje) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: why?

[the remark this answers is not in the document]
Comment (CKGrafico) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: ok after talk we are going to change arrow functions to bind(this), because the ideal case is one line arrow functions but are not allowed by linting. @borgitas21 we can talk about this tomorrow.

We can talk in the future with @molant and @sarvaje to allow one line lambdas, but not important at this moment :)

How we understand what is better:
ok
```
this.compatCSS = new CompatCSS(context, (...params) => {
    this.testFeatureIsSupportedInBrowser(...params);
});
```

better
```
new CompatCSS(context, this.testFeatureIsSupportedInBrowser.bind(this));
```

best
```
new CompatCSS(context, (...x) => this.testFeatureIsSupportedInBrowser(...x));
```

[the remark this answers is not in the document]
Comment (sarvaje) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: IMO, we should go with bind even if we enable one line lambdas.

The effect is the same, we are creating a new function in both cases, but in the case of the `bind`, you don't care about the parameters or anything else.

At least for me, `bind` is cleaner than one line lambda.

@webhintio/contributors thoughts?

File: packages/hint-compat-api/src/compat-api-css-next.ts [hunk 26]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    private async onParseCSS(styleParse: StyleParse): Promise<void> {
+        const { resource } = styleParse;
+
+        this.compatCSS.setResource(resource);
+        await this.compatCSS.searchCSSFeatures(this.compatApi.compatDataApi, this.mdnBrowsersCollection, styleParse);
+    }
+
+    private async testFeatureIsSupportedInBrowser(browsersToSupport: BrowserSupportCollection, browserToSupportName: string, browserInfo: SupportStatement, featureName: string, prefix?: string, location?: ProblemLocation): Promise<void> {
```
Comment (sarvaje) [inline packages/hint-compat-api/src/compat-api-css-next.ts]: There are few methods with a lot of parameters (like this one).

What do you think about having a type "Feature" with all the data needed?

**Comment by borgitas21** [comment]
FYI: We are refactoring a bit the `compat-api-css.ts` and `compat-api-css-next.ts` classes since we have detected many duplicated code between them.

File: packages/hint-compat-api/src/compat-api-css-next.ts (hunk 26 above)
Comment (CKGrafico) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: We are thinking in something like
FeatureInfo (featureName: string, prefix?: string, location?: ProblemLocation)
BrowsersInfo (browsersToSupport: BrowserSupportCollection, browserToSupportName: string, browserInfo: SupportStatement)

File: packages/hint-compat-api/src/compat-api-css-next.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (borgitas21) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: Implemented as part of #14

File: packages/hint-compat-api/src/compat-api-css-next.ts (hunk 26 above)
Comment (borgitas21) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: Implemented as part of #14

Comment (sarvaje) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: as part of #14?

>We are thinking in something like
FeatureInfo (featureName: string, prefix?: string, location?: ProblemLocation)
BrowsersInfo (browsersToSupport: BrowserSupportCollection, browserToSupportName: string, browserInfo: SupportStatement)

Looks good to me

File: packages/hint-compat-api/src/compat-api-css-base.ts [hunk 27]
```diff
@@ -0,0 +1,113 @@
...
+import { CSSFeatureStatus } from './enums';
+
+/*
+ * ------------------------------------------------------------------------------
+ * Public
+ * ------------------------------------------------------------------------------
+ */
+
+export default abstract class BaseCompatApiCSS implements IHint {
+    public static readonly meta = meta;
```
Comment (sarvaje) [inline packages/hint-compat-api/src/compat-api-css-base.ts]: Is the metadata common?

File: packages/hint-compat-api/src/compat-api-css-base.ts [hunk 28]
```diff
@@ -0,0 +1,113 @@
...
+    abstract isVersionValueSupported(version: VersionValue): boolean;
+    abstract isVersionValueTestable(version: VersionValue): boolean;
+    abstract getStatusNameValue(): CSSFeatureStatus;
+
+    public constructor(context: HintContext<StyleEvents>, isCheckingNotBroadlySupported: boolean) {
+        this.mdnBrowsersCollection = userBrowsers.convert(context.targetedBrowsers);
+        this.compatApi = new CompatApi('css', this.mdnBrowsersCollection, isCheckingNotBroadlySupported);
+        this.compatCSS = new CompatCSS(context, this.testFeatureIsSupportedInBrowser.bind(this));
+
+        context.on('parse::end::css', async (styleParse: StyleParse) => {
```
Comment (sarvaje) [inline packages/hint-compat-api/src/compat-api-css-base.ts]: use `this.onParseCSS(styleParse).bind(this)` instead.

File: packages/hint-compat-api/src/compat-api-css-base.ts [hunk 29]
```diff
@@ -0,0 +1,113 @@
...
+
+    private async testFeatureIsSupportedInBrowser(browser: BrowsersInfo, feature: FeatureInfo): Promise<void> {
+        if (!this.compatApi.isBrowserToSupportPartOfBrowsersCollection(browser.browsersToSupport, browser.browserToSupportName)) {
+            return;
+        }
+
+        const browserFeatureSupported = this.compatApi.getSupportStatementFromInfo(browser.browserInfo, feature.prefix);
+
+        if (browserFeatureSupported) {
+            this.testVersionByBrowsers(browser, browserFeatureSupported, feature);
```
Comment (sarvaje) [inline packages/hint-compat-api/src/compat-api-css-base.ts]: `testVersionByBrowsers` is an async function, you need to wait for it to finish.
```suggestion
            await this.testVersionByBrowsers(browser, browserFeatureSupported, feature);
```

File: packages/hint-compat-api/src/compat-api-css-base.ts [hunk 30]
```diff
@@ -0,0 +1,113 @@
...
+        } else {
+            const message = `${feature.featureName} of CSS was never supported on any of your browsers to support.`;
+
+            await this.compatCSS.reportIfThereIsNoInformationAboutCompatibility(message, browser.browsersToSupport, browser.browserToSupportName, feature.featureName, feature.location);
+        }
+    }
+
+    private async testVersionByBrowsers(browser: BrowsersInfo, browserFeatureSupported: SimpleSupportStatement, feature: FeatureInfo) {
```
Comment (sarvaje) [inline packages/hint-compat-api/src/compat-api-css-base.ts]: I think we should keep always as much the order as possible:
1st. Browser Info
2nd. Feature Info
3rd other Info

What do you think?

File: packages/hint-compat-api/src/compat-api-css-next.ts [hunk 31]
```diff
@@ -0,0 +1,60 @@
...
+import BaseCompatApiCSS from './compat-api-css-base';
+import { CSSFeatureStatus } from './enums';
+
+/*
+ * ------------------------------------------------------------------------------
+ * Public
+ * ------------------------------------------------------------------------------
+ */
+
+/*
```
Comment (sarvaje) [inline packages/hint-compat-api/src/compat-api-css-next.ts]: remove commented code

File: packages/hint-compat-api/src/compat-api-css-next.ts [hunk 32]
```diff
@@ -0,0 +1,60 @@
...
+ * type UserPrefixes = {
+ *    [key: string]: boolean;
+ * };
+ */
+
+export default class extends BaseCompatApiCSS {
+    /* private userPrefixes: UserPrefixes = {}; */
```
Comment (sarvaje) [inline packages/hint-compat-api/src/compat-api-css-next.ts]: remove commented code

File: packages/hint-compat-api/src/compat-api-css-next.ts [hunk 33]
```diff
@@ -0,0 +1,60 @@
...
+
+    public isSupportedVersion(currentVersion: number, version: number) {
+        return version >= currentVersion;
+    }
+
+    public getStatusNameValue(): CSSFeatureStatus {
+        return CSSFeatureStatus.Added;
+    }
+/*
+ * private addUserUsedPrefixes(browserName: string, featureName: string): void {
```
Comment (sarvaje) [inline packages/hint-compat-api/src/compat-api-css-next.ts]: Remove commented code

File: packages/hint-compat-api/src/compat-api-css-base.ts (hunk 27 above)
Comment (borgitas21) [reply packages/hint-compat-api/src/compat-api-css-base.ts]: Yes, you are right!

File: packages/hint-compat-api/src/compat-api-css-base.ts (hunk 30 above)
Comment (borgitas21) [reply packages/hint-compat-api/src/compat-api-css-base.ts]: +1

File: packages/hint-compat-api/src/compat-api-css-base.ts (hunk 29 above)
Comment (borgitas21) [reply packages/hint-compat-api/src/compat-api-css-base.ts]: 👍

File: packages/hint-compat-api/src/compat-api-css-next.ts (hunk 26 above)
Comment (borgitas21) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: > as part of #14?
My mistake, I meant https://github.com/CKGrafico/hint/pull/14

File: packages/hint-compat-api/src/compat-api-css-next.ts (hunk 33 above)
Comment (borgitas21) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: Done

File: packages/hint-compat-api/src/compat-api-css-next.ts (hunk 32 above)
Comment (borgitas21) [reply packages/hint-compat-api/src/compat-api-css-next.ts]: Done

**Comment by CKGrafico** [comment]
@webhintio/contributors Is this ready to close? we are going start new feature

**Comment by molant** [comment]
@CKGrafico don't wait for us to start in the new hint. Hopefully we will get this review and merged today or tomorrow.

**Comment by CKGrafico** [comment]
Ok thanks @molant we are going to create a new PR (from this branch to a new one called `feature/hint-compat-api-html`) with @borgitas21 and request `status:work-in-progress`

**Comment by borgitas21** [comment]
@antross it sounds good to me grouping errors by feature. However, I'm not sure how to handle standard and prefixed reports. Could you provide me the output for the box-lines example? Thanks :)

**Comment by antross** [comment]
> However, I'm not sure how to handle standard and prefixed reports. Could you provide me the output for the box-lines example? Thanks :)

@borgitas21 I think I see the confusion as that specific line of code didn't appear in my screenshot. The only reference to `box-lines` in my source file was `-webkit-` prefixed. All of the `box-lines` reports were generated in reference to this line of code (whether the report referred to a `-webkit-` prefix or not):

```css
    -webkit-box-lines: single;
```

Had the code contained a separate, un-prefixed reference to `box-lines`, I think a separate message for that would still be reasonable. Perhaps the "standard" reports in the screenshot were a bug?

**Comment by borgitas21** [comment]
After our latest changes, the amount of reports should be reduced drastically. Please let us know your feedback.

**Comment by molant** [comment]
🎉🎉🎉

@antross can you please merge it then?

Thanks!


________________________________

**Comment by CKGrafico** [comment]
Amazing work everyone!
